trino/trino.go

```diff
@@ -1253,13 +1264,263 @@ type queryResponse struct {
+			if err != nil {
+				return nil, fmt.Errorf("failed to decode inline segment at index %d: %v", segmentIndex, err)
+			}
+
+			queryData = append(queryData, decodedData...)
+		case "spooled":
+			uri, ok := segment["uri"].(string)
+			if !ok || uri == "" {
+				return nil, fmt.Errorf("missing or invalid 'uri' field in spooled segment at index %d", segmentIndex)
+			}
+			ackUri, ok := segment["ackUri"].(string)
+			if !ok || ackUri == "" { // Is mandatory?
+				return nil, fmt.Errorf("missing or invalid 'ackUri' field in spooled segment at index %d", segmentIndex)
+			}
+			headers, ok := segment["headers"].(map[string]interface{}) // is mandatory?
+			if !ok {
+				return nil, fmt.Errorf("missing or invalid 'headers' field in spooled segment at index %d", segmentIndex)
+			}
+
+			data, err := sp.fetchSegment(uri, ackUri, headers)
+			if err != nil {
+				return nil, fmt.Errorf("failed to fetch segment from uri '%s' at index %d: %v", uri, segmentIndex, err)
+			}
+
+			decodedData, err := decode(data, sp.encoding, metadata)
+			if err != nil {
+				return nil, fmt.Errorf("failed to decode spooled segment at index %d: %v", segmentIndex, err)
+			}
+
+			queryData = append(queryData, decodedData...)
+		}
+
+	}
+
+	return queryData, nil
+}
+
+func (sp *spoolingProtocol) fetchSegment(uri, ackUri string, headers map[string]interface{}) ([]byte, error) {
+	req, err := http.NewRequestWithContext(sp.ctx, "GET", uri, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range headers {
+		headerSlice, ok := v.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("unsupported header type %T", v)
+		}
+
+		if len(headerSlice) > 1 {
+			return nil, fmt.Errorf("multiple values for header %s", k)
+		}
+
+		header, ok := headerSlice[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("unsupported header value type %T for spooled segment", headerSlice[0])
+		}
+		req.Header.Add(k, header)
+	}
+
+	resp, err := sp.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %d", resp.StatusCode)
+	}
+
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %v", err)
+	}
+
+	//acknowledge the segment read
+	go func() {
+		// TODO: handle ack erros
+		ackReq, err := http.NewRequestWithContext(sp.ctx, "GET", ackUri, nil)
+		if err != nil {
+			return
+		}
+
+		for k, values := range req.Header {
+			for _, v := range values {
+				ackReq.Header.Add(k, v)
+			}
+		}
+
+		resp, err := sp.httpClient.Do(ackReq)
+		if err != nil {
+			return
+		}
+		resp.Body.Close()
+	}()
+
+	return data, nil
+}
+
+func decode(data []byte, encoding string, metadata spoolingMetadata) ([]queryData, error) {
+	if metadata.UncompressedSize == 0 {
+		var queryDataList = make([]queryData, metadata.RowsCount)
+		decoder := json.NewDecoder(bytes.NewReader(data))
```

Comment:
We should not duplicate this block of code (json decoding). Extract decompression to a method, and call it first in this method.